File: src/test/com/intellij/lang/jsgraphql/resolve/GraphQLOperationsResolveTest.java
package com.intellij.lang.jsgraphql.resolve;

import com.intellij.lang.jsgraphql.GraphQLResolveTestCaseBase;
import com.intellij.lang.jsgraphql.psi.*;

public class GraphQLOperationsResolveTest extends GraphQLResolveTestCaseBase {

    private static final String GITHUB_SCHEMA = "GithubSchema.graphql";

    @Override
    protected String getBasePath() {
        return "/resolve/operations";
    }

    public void testQueryFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "name");
    }

    public void testCustomQueryFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "name");
    }

    public void testSelectionSetQueryFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "name");
    }

    public void testMutationFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "createUser");
    }

    public void testCustomMutationFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "createUser");
    }

    public void testSubscriptionFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "users");
    }

    public void testCustomSubscriptionFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "users");
    }

    public void testFragmentName() {
        doResolveWithOffsetTest(GraphQLFragmentDefinition.class, "fragment1");
    }

    public void testFragmentObjectField() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "name");
    }

    public void testFragmentObjectFieldExtension() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "email");
    }

    public void testFragmentInterfaceField() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "id");
    }

    public void testFragmentInterfaceFieldExtension() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "createdAt");
    }

    public void testFragmentUnionField() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "language");
    }

    public void testFragmentUnionFieldExtension() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "someOtherField");
    }

    public void testFragmentInlineAnonymous() {
        doResolveWithOffsetTest(GraphQLFieldDefinition.class, "id");
    }

    public void testFieldArgument() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "after");
    }

    public void testDirectiveArgument() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "second");
    }

    public void testDirective() {
        doResolveWithOffsetTest(GraphQLDirectiveDefinition.class, "someDir");
    }

    public void testInputValue() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "address");
    }

    public void testInputValueNested() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "zip");
    }

    public void testInputValueDefinitionDefault() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition.class, "street");
    }

    public void testGithubQueries() {
        myFixture.copyFileToProject(GITHUB_SCHEMA);
        doHighlightingTest();
    }

    public void testUnresolvedReferences() {
        myFixture.copyFileToProject(GITHUB_SCHEMA);
        doHighlightingTest();
    }
}

File: src/test/com/intellij/lang/jsgraphql/resolve/GraphQLOperationsResolveTest.kt
package com.intellij.lang.jsgraphql.resolve

import com.intellij.lang.jsgraphql.GraphQLResolveTestCaseBase
import com.intellij.lang.jsgraphql.psi.GraphQLDirectiveDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLFieldDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLFragmentDefinition
import com.intellij.lang.jsgraphql.psi.GraphQLInputValueDefinition

class GraphQLOperationsResolveTest : GraphQLResolveTestCaseBase() {

    companion object {
        private const val GITHUB_SCHEMA = "GithubSchema.graphql"
    }

    override fun getBasePath(): String {
        return "/resolve/operations"
    }

    fun testQueryFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "name")
    }

    fun testCustomQueryFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "name")
    }

    fun testSelectionSetQueryFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "name")
    }

    fun testMutationFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "createUser")
    }

    fun testCustomMutationFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "createUser")
    }

    fun testSubscriptionFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "users")
    }

    fun testCustomSubscriptionFieldRoot() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "users")
    }

    fun testFragmentName() {
        doResolveWithOffsetTest(GraphQLFragmentDefinition::class.java, "fragment1")
    }

    fun testFragmentObjectField() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "name")
    }

    fun testFragmentObjectFieldExtension() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "email")
    }

    fun testFragmentInterfaceField() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "id")
    }

    fun testFragmentInterfaceFieldExtension() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "createdAt")
    }

    fun testFragmentUnionField() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "language")
    }

    fun testFragmentUnionFieldExtension() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "someOtherField")
    }

    fun testFragmentInlineAnonymous() {
        doResolveWithOffsetTest(GraphQLFieldDefinition::class.java, "id")
    }

    fun testFieldArgument() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "after")
    }

    fun testDirectiveArgument() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "second")
    }

    fun testDirective() {
        doResolveWithOffsetTest(GraphQLDirectiveDefinition::class.java, "someDir")
    }

    fun testInputValue() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "address")
    }

    fun testInputValueNested() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "zip")
    }

    fun testInputValueDefinitionDefault() {
        doResolveWithOffsetTest(GraphQLInputValueDefinition::class.java, "street")
    }

    fun testGithubQueries() {
        myFixture.copyFileToProject(GITHUB_SCHEMA)
        doHighlightingTest()
    }

    fun testUnresolvedReferences() {
        myFixture.copyFileToProject(GITHUB_SCHEMA)
        doHighlightingTest()
    }

    fun testFragmentExplicitDocumentsGlob() {
        doProjectResolveTest(
            "bar/bar.js",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "bar/schema.graphql"
        )
    }

    fun testFragmentFallbackToFirstNonStrictProject() {
        doProjectResolveTest(
            "bar/bar.js",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "foo/schema.graphql"
        )
    }

    fun testFragmentFallbackToFirstNonStrictProjectSkipInclude() {
        doProjectResolveTest(
            "bar/bar.js",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "bar/schema.graphql"
        )
    }

    fun testFragmentMatchedBySchema() {
        doProjectResolveTest(
            "bar/bar.js",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "bar/schema.graphql"
        )
    }

    fun testFragmentInjectedInHtml() {
        doProjectResolveTest(
            "index.html",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "fragments.graphql"
        )
    }

    fun testFragmentInjectedResolvedToOtherInjection() {
        doProjectResolveTest(
            "index.html",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "fragments.js"
        )
    }

    fun testFragmentIncluded() {
        doProjectResolveTest(
            "query.graphql",
            GraphQLFragmentDefinition::class.java,
            "UserFragment",
            "fragments.graphql"
        )
    }

    fun testFragmentExcluded() {
        doProjectHighlighting("query.graphql")
    }
}
